packages/api/src/util/animal.js

```diff
@@ -0,0 +1,76 @@
+/*
+ *  Copyright 2024 LiteFarm.org
+ *  This file is part of LiteFarm.
+ *
+ *  LiteFarm is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  LiteFarm is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *  GNU General Public License for more details, see <https://www.gnu.org/licenses/>.
+ */
+
+import AnimalModel from '../models/animalModel.js';
+import AnimalBatchModel from '../models/animalBatchModel.js';
+
+/**
+ * Update the map with type counts.
+ * @param {Object} typeIdCountMap - The map object with typeId as keys.
+ * @param {Array<Object>} typeIdCountObjects - Array of objects containing type ids and counts.
+ * @param {string} typeIdColumnName - The name of the type id column - 'default_type_id' or 'custom_type_id'.
+ * @example
+ * // Example usage:
+ * const typeIdCountMap = { '1': 2 };
+ * const typeIdCountObjects = [
+ *   { default_type_id: 1, count: 5 },
+ *   { default_type_id: 2, count: 3 },
+ *   { default_type_id: 1, count: 2 }
+ * ];
+ * const typeIdColumnName = 'default_type_id';
+ * await updateMapWithTypeCounts(map, typeIdCountObjects, typeIdColumnName);
+ * console.log(typeIdCountMap);
+ *
+ * // Example output:
+ * // { '1': 9, '2': 3 }
+ */
+export const updateMapWithTypeCounts = async (
+  typeIdCountMap,
+  typeIdCountObjects,
+  typeIdColumnName,
+) => {
+  typeIdCountObjects.forEach((typeIdCountObj) => {
+    const typeId = typeIdCountObj[typeIdColumnName];
+    if (!typeIdCountMap[typeId]) {
+      typeIdCountMap[typeId] = 0;
+    }
+    typeIdCountMap[typeId] += +typeIdCountObj.count;
+  });
+};
+
+/**
+ * Retrieves data from animal and animal_batch tables and returns a map with type ids as keys and their corresponding counts.
+ * @param {string} farmId
+ * @param {string} typeIdColumnName - The name of the type id column - 'default_type_id' or 'custom_type_id'.
+ * @returns {Promise<Object>} A map with type ids as keys and their corresponding counts.
+ */
+export const getAnimalTypeIdCountMap = async (farmId, typeIdColumnName) => {
+  const animals = await AnimalModel.query()
+    .select(typeIdColumnName)
+    .count()
+    .groupBy(typeIdColumnName)
+    .where('farm_id', farmId)
+    .whereNotDeleted();
+  const animalBatches = await AnimalBatchModel.query()
+    .select(typeIdColumnName, 'count')
+    .where('farm_id', farmId)
+    .whereNotDeleted();
+
+  const map = {};
+  updateMapWithTypeCounts(map, animals, typeIdColumnName);
+  updateMapWithTypeCounts(map, animalBatches, typeIdColumnName);
+
+  return map;
+};
```

Comment:
You can simplify this by using `UNION` to get results from both tables, group them by your column name and return the counts

```sql
select default_type_id as type, count(*) from (
  (select default_type_id from animal a)
  union all
  (select default_type_id from animal_batch ab)
) as animal_batch_union
group by default_type_id
```

I think if you do it this way you'll get the counts for each type directly from the query and won't need any `updateMap` extra functions.

Reply:
Oh true, I forgot that what were returned from the two tables were the same structure.
Thank you!

Reply:
I might have missed some context here -- maybe from the animal identifier discussion? -- but why are we now using raw SQL in the controllers if it's more expedient?

I've always remembered this bit of technical guidance from our [Pre-integration peer review document](https://lite-farm.atlassian.net/wiki/spaces/LITEFARM/pages/1191575576/Pre-integration+peer+review) because it made a lot of intuitive sense to me

![Screenshot 2024-02-15 at 3 15 36 PM](https://github.com/LiteFarmOrg/LiteFarm/assets/82857964/3f665277-3090-4255-abdb-42db0452ba4c)

Are we revising these guidelines?

Reply:
You haven't missed anything, I just don't remember those rules, so I will read and follow the guidelines.
I wrote new queries in controllers. Are they supposed to be in models?

Reply:
Well it was more a question for @antsgar in suggesting an SQL solution (I thought the original map read pretty well)

Reply:
Sorry 😂 

Understood. I did those two approaches, and with the raw SQL approach, I was able to do everything in one query, but there were full of errors at first. I forgot to check `deleted` and `farm_id` and ended up adding more tests.

Reply:
Oh yeah, you're right the controller should call the model instead of executing the query and it's the model's responsability to query the database, but if we wanted a more complex query we could just add that to the model and call it from the controller -- my code suggestion didn't necessarily imply using that code directly in this file, my bad! That being said I understand why we went with this approach here because of the counts for batches, so it makes sense now

Reply:
Oh @SayakaOno I just saw you implemented the more complex query! Query looks great, nice job 🙌 but I think @kathyavini is right that it belongs in the model. I'd create a new method in the model and call it from the controller. Wdyt?

Reply:
Thank you Anto, sounds good. I will move the queries to the models.
But as Joyce was asking, I'm curious if we should avoid writing raw SQL and use ORM...

Reply:
If you can write the same query with Knex then sure, since we're already using Knex that should give us a more readable output! Sometimes Knex has some limitations and some queries need to be written raw, but if both approaches are doable I'd agree the ORM should be preferred if we do have an ORM like in this case

Reply:
I understand.
I believe those queries are some of those that need to be written raw.
I have moved the queries to the models. It would be appreciated if you could re-review when you have the time. Thank you!